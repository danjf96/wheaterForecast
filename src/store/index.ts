
import { applyMiddleware, combineReducers, createStore } from 'redux'
import ReduxThunk from 'redux-thunk';

const reducers = {

}

const combinedReducers =  combineReducers(reducers)

const store = createStore(combinedReducers, {}, applyMiddleware(ReduxThunk));

// We can use RootState type in every file in project
declare global {
    type RootState = ReturnType<typeof store.getState>
}

// Thanks to that you will have ability to use useSelector hook with state value
declare module 'react-redux' {
    interface DefaultRootState extends RootState { }
}

export default store

