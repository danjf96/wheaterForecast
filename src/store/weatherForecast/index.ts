import { STATEWEATHERFORECAST, CURREMTWEATHER, Types } from "./types";
import geolocation from 'react-native-geolocation-service'
import api from "../../services/api";


//REDUCERS
const INITIAL_STATE: STATEWEATHERFORECAST = {
    loading: true,
    base: "test",
    clouds: {all: 0},
    cod: 0,
    coord: {lon: -122.4064, lat: 37.7858},
    dt: 0,
    id: 0,
    main: {
        feels_like: 0,
        humidity: 0,
        pressure: 0,
        temp: 0,
        temp_max: 0,
        temp_min: 0
    },
    name: "teste",
    sys: { type: 0, id: 0, country: 'teste', sunrise: 0, sunset: 0},
    timezone: -25200,
    visibility: 10000,
    weather: [],
    wind: {speed: 0, deg: 0, gust: 0}
}

//STATE CASES
const STATE = (state = INITIAL_STATE, action: any):STATEWEATHERFORECAST => {
    switch (action.type) {
        case Types.CHANGE:
            return { ...state, ...action.payload }
        case Types.LOAD:
            return { ...state, loading: action.loading }
        case Types.CLEAN:
            return INITIAL_STATE
        default:
            return state;
    }
}

//Actions Creators
export const changeWheatherForecast = (payload:any, key?: string) => {
    return async (dispatch: (arg0:any) => any) => {
        dispatch({ type: Types.CHANGE, payload: key ? { [key]: payload } : payload })
    }

}

export const getWheatherForecast = (coords: geolocation.GeoCoordinates) => {
    return async (dispatch: (arg0:any) => any) => {
        dispatch({ type: Types.LOAD, loading: true })
       try{
            const { data }:{ data: CURREMTWEATHER} = await api.get(`data/2.5/weather?lat=${coords.latitude}&lon=${coords.longitude}&lang=pt_br&units=metric`)
            dispatch({ type: Types.CHANGE, payload: { ...data, loading: false} })
       }catch(e){
            dispatch({ type: Types.LOAD, loading: false })
       }
    }

}

export default STATE