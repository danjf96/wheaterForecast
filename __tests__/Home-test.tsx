/**
 * @format
 */

 import 'react-native';
 import React from 'react';
 import { act, fireEvent, render, waitFor } from 'react-native-testing-library';
 import Home from '../src/screens/home';
 import store from '../src/store'
 import { Provider } from 'react-redux';
 jest.mock('react-native/Libraries/EventEmitter/NativeEventEmitter');

 const component = (
   <Provider store={store} > <Home navigation={{ navigate: (screen:'string') => null }}/> </Provider>
 );
 
 it('renders correctly', async () => {
   const screen = render(component)
   expect(screen.toJSON()).toMatchSnapshot()
 });
 
 describe('test loading and render items or not', () => {
 
   test('renders correctly loading', async () => {
 
     const { getByTestId, getAllByTestId } = render( component )   
     const weatherForecast = store.getState().weatherForecast
     expect(weatherForecast.location).toBe(undefined)
    //  expect(weatherForecast.loading).toBeTruthy()
 
   });
 
   test('renders correctly location', async () => {
 
        const { getAllByTestId, getByTestId, getByText, update } = await waitFor(() =>
            render(component)
        ); 

        await waitFor( () => {
            const initialMain =  {
                feels_like: 0,
                humidity: 0,
                pressure: 0,
                temp: 0,
                temp_max: 0,
                temp_min: 0
            }
            const weatherForecast = store.getState().weatherForecast
            expect(weatherForecast.main).not.toBe(initialMain) 
        })
 
   });
   
 }) 
   