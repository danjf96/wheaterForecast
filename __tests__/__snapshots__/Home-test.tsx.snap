// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
Array [
  " ",
  <BVLinearGradient
    colors={
      Array [
        4278224639,
        4288499932,
      ]
    }
    endPoint={
      Object {
        "x": 0.5,
        "y": 1,
      }
    }
    locations={null}
    startPoint={
      Object {
        "x": 0.5,
        "y": 0,
      }
    }
    style={
      Array [
        Object {
          "flex": 1,
        },
        undefined,
      ]
    }
  >
    <RCTScrollView
      refreshControl={
        <RefreshControlMock
          onRefresh={[Function]}
          refreshing={false}
        />
      }
      style={
        Object {
          "backgroundColor": "transparent",
          "flex": 1,
          "minHeight": 1334,
          "width": "100%",
        }
      }
    >
      <RCTRefreshControl />
      <View>
        <RCTSafeAreaView
          emulateUnlessSupported={true}
        >
          <View
            style={
              Object {
                "backgroundColor": "transparent",
                "marginTop": 30,
                "width": "100%",
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "justifyContent": "center",
                  "marginBottom": 20,
                  "width": "100%",
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "white",
                    "fontSize": 20,
                    "textAlign": "center",
                  }
                }
              >
                 - 
              </Text>
              <View
                accessible={true}
                collapsable={false}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "opacity": 1,
                  }
                }
                testID="reload"
              >
                <Image
                  source={
                    Object {
                      "testUri": "../../../src/assets/images/reload.png",
                    }
                  }
                  style={
                    Object {
                      "height": 48,
                      "marginLeft": 15,
                      "resizeMode": "contain",
                      "width": 48,
                    }
                  }
                />
              </View>
            </View>
            <View
              style={
                Object {
                  "alignItems": "center",
                  "flexDirection": "column",
                  "width": "100%",
                }
              }
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                  }
                }
              >
                <View
                  style={
                    Object {
                      "alignItems": "center",
                      "flexDirection": "column",
                      "justifyContent": "center",
                    }
                  }
                >
                  <Text
                    style={
                      Object {
                        "color": "white",
                        "fontSize": 60,
                        "fontWeight": "bold",
                        "marginBottom": 15,
                      }
                    }
                  >
                    0.0
                     °C
                  </Text>
                  <Text
                    style={
                      Object {
                        "color": "white",
                        "fontSize": 18,
                        "fontWeight": "bold",
                        "textTransform": "capitalize",
                      }
                    }
                  />
                  <View
                    style={
                      Object {
                        "flexDirection": "row",
                        "marginTop": 15,
                      }
                    }
                  >
                    <Text
                      style={
                        Object {
                          "color": "white",
                          "fontSize": 15,
                          "marginRight": 10,
                        }
                      }
                    >
                      Máx: 
                      0
                       °C
                    </Text>
                    <Text
                      style={
                        Array [
                          Object {
                            "color": "white",
                            "fontSize": 15,
                            "marginRight": 10,
                          },
                          Object {
                            "marginRight": 0,
                          },
                        ]
                      }
                    >
                      min: 
                      0
                       °C
                    </Text>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </RCTSafeAreaView>
      </View>
    </RCTScrollView>
  </BVLinearGradient>,
  " ",
]
`;
